packages/key-did-provider-webcrypto/src/index.ts

```diff
@@ -0,0 +1,188 @@
+ * This is a DID Provider which implements [EIP2844](https://eips.ethereum.org/EIPS/eip-2844) for `did:key:` using webcrypto with non-extractable keys.
+ *
+ * ## Installation
+ *
+ * ```
+ * npm install --save @didtools/key-webcrypto
+ * ```
+ *
+ * ## Usage
+ *
+ * ```js
+ * import { CryptoKeyProvider, generateP256KeyPair } from '@didtools/key-webcrypto'
+ * import KeyResolver from 'key-did-resolver'
+ * import { DID } from 'dids'
+ *
+ * const keyPair = generateP256KeyPair()
+ * const provider = new WebcryptoProvider(keypair)
+ * const did = new DID({ provider, resolver: KeyResolver.getResolver() })
+ * await did.authenticate()
+ *
+ * // log the DID
+ * console.log(did.id)
+ *
+ * // create JWS
+ * const { jws, linkedBlock } = await did.createDagJWS({ hello: 'world' })
+ *
+ * // verify JWS
+ * await did.verifyJWS(jws)
+ * ```
+ *
+ * @module @didtools/key-webcrypto
+ */
+
+import type {
+  AuthParams,
+  CreateJWSParams,
+  DIDMethodName,
+  DIDProviderMethods,
+  DIDProvider,
+  GeneralJWS,
+} from 'dids'
+import stringify from 'fast-json-stable-stringify'
+import { RPCError, createHandler } from 'rpc-utils'
+import type { HandlerMethods, RPCRequest, RPCResponse, SendRequestFunc } from 'rpc-utils'
+import * as u8a from 'uint8arrays'
+import varint from 'varint'
+
+function toStableObject(obj: Record<string, any>): Record<string, any> {
+  return JSON.parse(stringify(obj)) as Record<string, any>
+}
+
+/**
+ *  compress a public key with points x,y expressed as UintArrays
+ * source: https://stackoverflow.com/questions/17171542/algorithm-for-elliptic-curve-point-compression
+ *
+ *  @param {Uint8Array}  x  x point of public key
+ *  @param {Uint8Array}  y  y point of public key
+ *  @return {Uint8Array} compressed form of public key as Uint8Array
+ */
+export function ecPointCompress(x: Uint8Array, y: Uint8Array): Uint8Array {
+  const out = new Uint8Array(x.length + 1)
+
+  out[0] = 2 + (y[y.length - 1] & 1)
+  out.set(x, 1)
+
+  return out
+}
+
+// export the raw public key from the CryptoKeyPair using the webcrypto api
+export async function getPublicKey({ publicKey }: CryptoKeyPair): Promise<Uint8Array> {
+  const rawKey = await window.crypto.subtle.exportKey('raw', publicKey)
+  // convert raw key with x,y to a compressed key
+  const compressedKey = ecPointCompress(
+    new Uint8Array(rawKey.slice(1, 33)),
+    new Uint8Array(rawKey.slice(33, 65))
+  )
+  return compressedKey
+}
+
+export function encodeDIDFromPub(publicKey: Uint8Array): string {
+  const CODE = varint.encode(0x1200) // p-256 multicodec
+  const bytes = u8a.concat([CODE, publicKey])
+  return `did:key:z${u8a.toString(bytes, 'base58btc')}`
+}
+
+// A method that generates a new key using webcrypto, p-256, extractable false.
+export async function generateP256KeyPair(): Promise<CryptoKeyPair> {
+  const { privateKey, publicKey } = await window.crypto.subtle.generateKey(
+    {
+      name: 'ECDSA',
+      namedCurve: 'P-256',
+    },
+    false, // not extractable
```

Comment:
The reason I implemented this library is so that it can be used as default in `did-session`. This would remove the need for devs to figure out where to store the session and manage the session key, because this non-extractable key can be persisted in indexeddb, along with the CACAO.

I think it could make sense to add additional options for this parameter and the curve used as well, but don't want to expand the scope here too much.

Reply:
Makes sense, thanks. We can always add more options over time as needed.